src/sparseml/pytorch/optim/modifier_distillation.py

```diff
@@ -0,0 +1,326 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Modifier for performing model distillation
+"""
+
+
+from typing import Any, Dict, Iterable, List, Optional
+
+import torch.nn.functional as TF
+from torch import Tensor
+from torch.nn import Module
+from torch.optim import Optimizer
+
+from sparseml.optim import ModifierProp
+from sparseml.pytorch.optim.modifier import PyTorchModifierYAML, ScheduledModifier
+from sparseml.pytorch.utils import BaseLogger, tensors_module_forward
+
+
+__all__ = [
+    "DistillationModifier",
+]
+
+
+@PyTorchModifierYAML()
+class DistillationModifier(ScheduledModifier):
+    """
+    Adds a knowledge distillation loss based on a teacher model during the
+    loss_update phase of the SparseML lifecycle. A distillation_teacher
+    module must be provided as a kwarg to the Manager initialization and
+    loss_update(loss) must be called before any backwards pass in the integrated
+    training flow
+
+    | Sample yaml:
+    |   !DistillationModifier
```

Comment:
@bfineran just for the logability and tracability of our recipes it would be useful to require the model path for the teacher.

Reply:
Model path is local state that would break the recipes for anything outside of the environment it was originally used on. State for how to define the teacher should be managed externally to this with the caller code and passed in so we can keep this extendable and usable across systems

Reply:
To clarify a bit more, the integration within hugging face would be responsible for receiving the path to the teacher from the user and passing that into the modifiers

Reply:
resolving this - we should make a point to log the teacher model in integration scripts when possible to your point @spacemanidol